src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py

```diff
@@ -23,47 +29,241 @@ def _graph_client_factory(**_):
     return client
 
+def list_role_definitions(name=None, resource_group_name=None, resource_id=None,
+                          custom_role_only=False):
+    definitions_client = _auth_client_factory().role_definitions
+    scope = _build_role_scope(resource_group_name, resource_id,
+                              definitions_client.config.subscription_id)
+    return _search_role_definitions(definitions_client, name, scope, custom_role_only)
+
+def create_role_definition(role_definition):
+    role_id = uuid.uuid4()
+    if os.path.exists(role_definition):
+        role_definition = get_file_json(role_definition)
+    else:
+        role_definition = json.loads(role_definition)
+
+    #to workaround service defects, ensure property names are camel case
+    names = [p for p in role_definition if p[:1].isupper()]
+    for n in names:
+        new_name = n[:1].lower() + n[1:]
+        role_definition[new_name] = role_definition.pop(n)
+
+    if not 'name' in role_definition:
+        raise CLIError("please provide 'name'")
+    if not 'assignableScopes' in role_definition:
+        raise CLIError("please provide 'assignableScopes'")
+
+    permission = Permission(actions=role_definition.get('actions', None),
+                            not_actions=role_definition.get('notActions', None))
+    properties = RoleDefinitionProperties(role_name=role_definition['name'],
+                                          description=role_definition.get('description', None),
+                                          type=_CUSTOM_RULE,
+                                          assignable_scopes=role_definition['assignableScopes'],
+                                          permissions=[permission])
+
+    definition = RoleDefinition(name=role_id, properties=properties)
+
+    definitions_client = _auth_client_factory().role_definitions
+    return definitions_client.create_or_update(role_definition_id=role_id,
+                                               scope=properties.assignable_scopes[0],
+                                               role_definition=definition)
+
+
+def delete_role_definition(name, resource_group_name=None, resource_id=None,
+                           custom_role_only=False):
+    definitions_client = _auth_client_factory().role_definitions
+    scope = _build_role_scope(resource_group_name, resource_id,
+                              definitions_client.config.subscription_id)
+    roles = _search_role_definitions(definitions_client, name, scope, custom_role_only)
+    for r in roles:
+        definitions_client.delete(role_definition_id=r.name, scope=scope)
+
+def _search_role_definitions(definitions_client, name, scope, custom_role_only=False):
+    roles = definitions_client.list(scope, filter="roleName eq '{}'".format(name) if name else None)
+    if custom_role_only:
+        roles = [r for r in roles if r.properties.type == _CUSTOM_RULE]
+    return roles
+
+def show_role_create_template():
```

Comment:
After our discussion I understand the rationale. Once you've changed the create_role command to be multiple parameters, I'm not sure this would be necessary. Even if we preserve the ability to use a JSON template as the input, they could just get the template from the examples portion of the help file. I'd ask others what they think.

---
In reply to: [73603847](https://github.com/Azure/azure-cli/pull/602#discussion_r73603847) [](ancestors = 73603847)